content/guides/typescript-for-solid.mdx

```diff
@@ -0,0 +1,586 @@
+
+### Components
+
+#### The basics
+
+By default, components in Solid use the generic `Component<P>` type, where `P` represents the prop's type that is typically an object.
+
+```typescript
+import type { JSX, Component } from "solid-js";
+
+type Component<P = {}> = (props: P) => JSX.Element;
+```
+
+A `JSX.Element` denotes anything renderable by Solid, which could be a DOM node, array of JSX elements, or functions yielding `JSX.Element`.
+
+For a simple example, here's a counter:
+
+```typescript
+const Counter: Component = () => {
+  const [count, setCount] = createSignal(0);
+  return <button onClick={() => setCount((c) => c + 1)}>{count()}</button>;
+};
+```
+
+Trying to pass unneccessary props or children will result in type errors:
+
+```typescript
+<Counter />;  // ✔️
+<Counter initial={5} />;  // ❌: No 'initial' prop defined
+<Counter>hi</Counter>;  // ❌: Children aren't expected
+```
+
+#### Components with props
+
+For components that require the use of props, they can be typed using [generics](https://www.typescriptlang.org/docs/handbook/2/generics.html#handbook-content):
+
+```typescript
+const InitCounter: Component<{ initial: number }> = (props) => {
+  const [count, setCount] = createSignal(props.initial);
+  return <button onClick={() => setCount((c) => c + 1)}>{count()}</button>;
+};
+
+<InitCounter initial={5} />;
+```
+
+#### Components with children
+
+Often, components may need to accept child elements.
+For this, Solid provides `ParentComponent` that includes `children?` as an optional prop.
+If defining a component with the `function` keyword, `ParentProps` can be used as a helper for the props:
+
+```typescript
+import { JSX, ParentComponent, ParentProps } from "solid-js";
+
+type ParentProps<P = {}> = P & { children?: JSX.Element };
+
+const CustomCounter: ParentComponent = (props) => {
+  const [count, setCount] = createSignal(0);
+  return (
+    <button onClick={() => setCount((c) => c + 1)}>
+      {count()}
+      {props.children}
+    </button>
+  );
+};
+```
+
+In this example, `props` is inferred to be of the type `{children?: JSX.Element }`, streamlining the process of defining components that can accept children.
+
+##### Special component types
+
+Solid offers subtypes for components dealing uniquely with children:
+
+* **VoidComponent:** when a component should not accept children.
+* **FlowComponent:** designed for components like [`<Show>`](/reference/Show) or [`<For>`](/reference/For), typically requiring children and, occasionally, specific children types.
+
+These types make sure that the children are present and fit the required type, maintaining consistent component behaviour.
+
+### Event handling
+
+#### Basics
+
+The `JSX.EventHandler<T, E>` type allows you to define event handlers for a specific DOM element type `T` and event type `E.` 
+This provides the event handler function with events of the right type and makes sure it interacts correctly with elements of the expected type.
```

Comment:
This reads awkwardly to me. I'd like to think that the important thing here is that it is exactly[^1] the type used by the `onInput` property, but that isn't what I get when I read this. 

[^1]: Technically it's `EventHandlerUnion` but no one likes `BoundEventHandler`s.

Reply:
```suggestion
The `JSX.EventHandler<T, E>` type allows you to define event handlers for a specific DOM element.
This type ensures event handlers cater to a specific DOM element type `T` and event type `E`.
By providing the event handler function with events of the right type, it ensures that the event handler processes events of the correct type and interacts with elements as anticipated
```

Reply:
Thoughts?

Reply:
It's better, but it still reads like "this type will fit into the event handler, but something about it is different from the actual type of the event handler that will help me make fewer errors" rather than "this type is the event handler, and I can treat it as such", since the phrasing seems to avoid saying the latter which implies the former. 

I'm not sure if this is intentional to avoid confusion due to the nature of the actual type (which accepts both the event handler and bound event handler), but in that case I would still want to avoid implying that this event handler type is somehow special.